File: app/src/main/java/com/alexkn/syntact/domain/worker/FetchPhrasesWorker.java
package com.alexkn.syntact.domain.worker;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import com.alexkn.syntact.app.ApplicationComponentProvider;
import com.alexkn.syntact.app.Property;
import com.alexkn.syntact.data.common.AppDatabase;
import com.alexkn.syntact.data.dao.ClueDao;
import com.alexkn.syntact.data.dao.SolvableItemDao;
import com.alexkn.syntact.restservice.SolvableItemService;
import com.alexkn.syntact.restservice.SyntactService;

import javax.inject.Inject;

public class FetchPhrasesWorker extends Worker {

    private static final String TAG = FetchPhrasesWorker.class.getSimpleName();

    private final SolvableItemDao solvableItemDao;

    private final ClueDao clueDao;

    @Inject
    SolvableItemService solvableItemService;

    @Inject
    SyntactService syntactService;

    @Inject
    Property property;

    public FetchPhrasesWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {

        super(context, workerParams);
        ((ApplicationComponentProvider) context).getApplicationComponent().inject(this);

        AppDatabase database = AppDatabase.getDatabase(context);
        solvableItemDao = database.solvableItemDao();
        clueDao = database.clueDao();
    }

    @NonNull
    @Override
    public Result doWork() {

        //TODO do only if items on phone <= itemcount in bucket
        //TODO regularly update bucket phrase count
//        Data inputData = getInputData();
//        long bucketId = inputData.getLong("bucketId", -1L);
//        long timestamp = inputData.getLong("timestamp", Instant.now().toEpochMilli());
//        Instant now = Instant.ofEpochMilli(timestamp).truncatedTo(ChronoUnit.MINUTES);
//        if (bucketId == -1L) {
//            return Result.failure();
//        }
//
//        List<SolvableTranslationCto> solvableTranslationCtos = solvableItemService.fetchNewTranslations(bucketId, now, 10);
//        solvableTranslationCtos.forEach(cto -> {
//            solvableItemDao.insert(cto.getSolvableItem());
//            clueDao.insert(cto.getClue());
//        });

        return Result.success();
    }
}

File: app/src/main/java/com/alexkn/syntact/domain/worker/FetchPhrasesWorker.kt
package com.alexkn.syntact.domain.worker

import android.content.Context
import android.util.Log
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import com.alexkn.syntact.app.ApplicationComponentProvider
import com.alexkn.syntact.app.Property
import com.alexkn.syntact.app.TAG
import com.alexkn.syntact.data.dao.BucketDao
import com.alexkn.syntact.data.dao.ClueDao
import com.alexkn.syntact.data.dao.SolvableItemDao
import com.alexkn.syntact.data.model.Clue
import com.alexkn.syntact.data.model.SolvableItem
import com.alexkn.syntact.restservice.SyntactService
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import java.lang.Exception
import java.util.*
import javax.inject.Inject

class FetchPhrasesWorker(context: Context, workerParams: WorkerParameters) : CoroutineWorker(context, workerParams) {

    @Inject
    lateinit var syntactService: SyntactService

    @Inject
    lateinit var property: Property

    @Inject
    lateinit var clueDao: ClueDao

    @Inject
    lateinit var solvableItemDao: SolvableItemDao

    @Inject
    lateinit var bucketDao: BucketDao

    init {
        (context as ApplicationComponentProvider).applicationComponent.inject(this)
    }

    override suspend fun doWork(): Result = coroutineScope {

        val bucketId = inputData.getLong("bucketId", -1L)
        if (bucketId == -1L) {
            Result.failure()
        }

        val token = "Token " + property["api-auth-token"]

        val solvableItems = solvableItemDao.findSolvableItemsWithoutTranslation(bucketId)

        solvableItems.forEach {
            async {
                val translations = syntactService.getTranslations(token, it.translationUrl, Locale.getDefault().language)

                if (translations.size > 1) {
                    Log.i(TAG, "Multiple Translations not yet supported, using first translation and discarding others")
                }
                if (translations.isEmpty()) {
                    throw Exception("No Translation found for $it")
                }
                val translation = translations[0]
                val clue = Clue(
                        id = translation.id,
                        text = translation.text.capitalize(),
                        solvableItemId = it.id
                )
                clueDao.insert(clue)
                Log.i(TAG, "Saved Clue $clue")
            }
        }


        Result.success()
    }


}
